Auth: Enable retries and transaction for some db calls for auth tokens

bergquist opened the pull request:
ref https://github.com/grafana/grafana/issues/16638

bergquist committed:
auth: use the WithSession wrapper from sqlstore

the WithSession wrapper handles retries and connection
management so the caller dont have to worry about it.

bergquist committed:
auth: pass context from requests down to db layer

bergquist committed:
auth: use dbsession for token cleanup

bergquist commented:
Migrating from the old way to using `WithDbSession` I realized how bad the API is. We need to figure out a better API for this. Its very easy to write code like this
```go
var affected int64
err := srv.SQLStore.WithDbSession(ctx, func(dbSession *sqlstore.DBSession) error {
	// code

	affected, err := res.RowsAffected() 
	return err
})
```
and create an new variable in that scope instead of assigning the original one.

marefr commented on pkg/services/auth/auth_token.go:
```diff
@@ -80,14 +96,27 @@ func (s *UserAuthTokenService) CreateToken(userId int64, clientIP, userAgent str
-	exists, err := s.SQLStore.NewSession().Where("(auth_token = ? OR prev_auth_token = ?) AND created_at > ? AND rotated_at > ?", hashedToken, hashedToken, s.createdAfterParam(), s.rotatedAfterParam()).Get(&model)
+	var exists bool
+	var err error
+	err = s.SQLStore.WithDbSession(context.Background(), func(dbSession *sqlstore.DBSession) error {
```
s/context.Background()/ctx

marefr commented on pkg/services/auth/auth_token.go:
```diff
@@ -65,7 +76,12 @@ func (s *UserAuthTokenService) CreateToken(userId int64, clientIP, userAgent str
 	}
-	_, err = s.SQLStore.NewSession().Insert(&userAuthToken)
+
+	err = s.SQLStore.WithDbSession(context.Background(), func(dbSession *sqlstore.DBSession) error {
```
s/context.Background()/ctx

marefr commented on pkg/services/auth/auth_token.go:
```diff
@@ -35,14 +36,24 @@ func (s *UserAuthTokenService) Init() error {
+
+	var count int64
+	var err error
+	err = s.SQLStore.WithDbSession(context.Background(), func(dbSession *sqlstore.DBSession) error {
```
s/context.Background()/ctx

marefr commented on pkg/services/auth/auth_token.go:
```diff
@@ -100,7 +129,18 @@ func (s *UserAuthTokenService) LookupToken(unhashedToken string) (*models.UserTo
-		affectedRows, err := s.SQLStore.NewSession().Where("id = ? AND prev_auth_token = ? AND rotated_at < ?", modelCopy.Id, modelCopy.PrevAuthToken, expireBefore).AllCols().Update(&modelCopy)
+
+		var affectedRows int64
+		err = s.SQLStore.WithDbSession(context.Background(), func(dbSession *sqlstore.DBSession) error {
```
s/context.Background()/ctx

marefr commented on pkg/services/auth/auth_token.go:
```diff
@@ -116,7 +156,17 @@ func (s *UserAuthTokenService) LookupToken(unhashedToken string) (*models.UserTo
-		affectedRows, err := s.SQLStore.NewSession().Where("id = ? AND auth_token = ?", modelCopy.Id, modelCopy.AuthToken).AllCols().Update(&modelCopy)
+
+		var affectedRows int64
+		err = s.SQLStore.WithDbSession(context.Background(), func(dbSession *sqlstore.DBSession) error {
```
s/context.Background()/ctx

marefr commented on pkg/services/auth/auth_token.go:
```diff
@@ -183,12 +233,21 @@ func (s *UserAuthTokenService) TryRotateToken(token *models.UserToken, clientIP,
 
-	res, err := s.SQLStore.NewSession().Exec(sql, userAgent, clientIP, s.SQLStore.Dialect.BooleanStr(true), hashedToken, s.SQLStore.Dialect.BooleanStr(false), now.Unix(), model.Id, s.SQLStore.Dialect.BooleanStr(true), now.Add(-30*time.Second).Unix())
+	var affected int64
+	err = s.SQLStore.WithDbSession(context.Background(), func(dbSession *sqlstore.DBSession) error {
```
s/context.Background()/ctx

marefr commented on pkg/services/auth/auth_token.go:
```diff
@@ -199,14 +258,20 @@ func (s *UserAuthTokenService) TryRotateToken(token *models.UserToken, clientIP,
-	rowsAffected, err := s.SQLStore.NewSession().Delete(model)
+	var rowsAffected int64
+	var err error
+	err = s.SQLStore.WithDbSession(context.Background(), func(dbSession *sqlstore.DBSession) error {
```
s/context.Background()/ctx

marefr commented on pkg/services/auth/auth_token.go:
```diff
@@ -221,55 +286,71 @@ func (s *UserAuthTokenService) RevokeToken(token *models.UserToken) error {
-		return err
-	}
+func (s *UserAuthTokenService) RevokeAllUserTokens(ctx context.Context, userId int64) error {
+	return s.SQLStore.WithDbSession(context.Background(), func(dbSession *sqlstore.DBSession) error {
```
s/context.Background()/ctx

marefr commented on pkg/services/auth/auth_token.go:
```diff
@@ -221,55 +286,71 @@ func (s *UserAuthTokenService) RevokeToken(token *models.UserToken) error {
 
 	var result models.UserToken
-	token.toUserToken(&result)
+	err := s.SQLStore.WithDbSession(context.Background(), func(dbSession *sqlstore.DBSession) error {
```
s/context.Background()/ctx

marefr commented on pkg/services/auth/auth_token.go:
```diff
@@ -221,55 +286,71 @@ func (s *UserAuthTokenService) RevokeToken(token *models.UserToken) error {
-		token.toUserToken(&userToken)
-		result = append(result, &userToken)
-	}
+	err := s.SQLStore.WithDbSession(context.Background(), func(dbSession *sqlstore.DBSession) error {
```
s/context.Background()/ctx

marefr reviewed (changes requested):
Just to be clear this doesn't add any retries for sqlite right?

You did a very thorough work getting context in everywhere, but in auth_token.go you forget to use the provided ctx as argument to `WithDbSession` :)

torkelo commented:
Just to understand the db locked issue and how this solves it. Was it caused, you suspect, by opening & not closing sessions? 

Wonder if it would be possible to open a session in the middleware at the start and store it in the Request context, and the pass context down to sql layer and have the withDbSession reuse session in the context object. And have the middleware close the session on request completion.

bergquist committed:
use the context

bergquist commented:
Fixed the `context.Background` issues and switched to using transactions (inc retries) for rotation. I was considering adding retries for all WithDbSession requests as well but Im not sure about it. It could potentially create more requests to the db if it's exhausted. 

@torkelo yesterday I didn't like that idea but I might have changed my mind. We still need some way of managing session background tasks. Let's discuss that in another issue. I think we need to improve this soon as more people will be working in the backend.

bergquist committed:
use transaction when updating current token

marefr commented on pkg/services/quota/quota.go:
```diff
@@ -20,6 +22,17 @@ func (qs *QuotaService) Init() error {
+	fmt.Println("kasdgfkjhgsdafkjhgasdkjhf")
+	fmt.Println("kasdgfkjhgsdafkjhgasdkjhf")
+	fmt.Println("kasdgfkjhgsdafkjhgasdkjhf")
+
```
Please remove this :smile:

bergquist replied:
haha

marefr reviewed (changes requested):
Everything looks good besides some dummy code that seems to have been sneaked in, see comment.

bergquist committed:
removes debug print out

marefr reviewed (approved):
:tada:
